0000-deb-repo-additional-headers.md

```diff
@@ -0,0 +1,166 @@
+```
+Package: apache2-data
+Architecture: all
+Version: 2.4.29-1ubuntu4.6
+Multi-Arch: foreign
+Priority: optional
+Section: httpd
+Source: apache2
+Origin: Ubuntu
+Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
+Original-Maintainer: Debian Apache Maintainers <debian-apache@lists.debian.org>
+Bugs: https://bugs.launchpad.net/ubuntu/+filebug
+Installed-Size: 870
+Filename: pool/main/a/apache2/apache2-data_2.4.29-1ubuntu4.6_all.deb
+Size: 159984
+MD5sum: 7d4e4ad49eb6c5014d5453721573e637
+SHA1: 1824d3d176e7db8010216abc2f0962ac2fe5531f
+SHA256: 594070814770a028a9eb3924d02b22fcb818d83b311710aaac304a9064479292
+Homepage: http://httpd.apache.org/
+Description: Apache HTTP Server (common files)
+Task: lamp-server
+Description-md5: 9f2fab36019a61312dec627d1cd80365
+Supported: 5y
+```
+
+The database schema needs to be extended with a new table to store the additional fields not covered by the existing `rhnPackage*` tables:
+```
+rhnPackageExtraTagKey
+=====================
+id          NUMERIC NOT NULL
+name        VARCHAR(256) NOT NULL
+
+rhnPackageExtraTag
+====================
+package_id  NUMERIC NOT NULL
+                   REFERENCES rhnPackage (id),
+key_id      NUMERIC NOT NULL
+                   REFERENCES rhnPackageExtraTagKey (id),
+value       VARCHAR(2048) NOT NULL,
+primary key (package_id, key_id)
+```
+
+The class `com.redhat.rhn.taskomatic.task.repomd.DebPackageWriter` must be enhanced to write any additional fields stored in the database into the generated `Packages.gz` file.
+
+_Note_: The same structure could be used to store additional RPM headers not yet stored. The corresponding class that needs to be changed is `com.redhat.rhn.taskomatic.task.repomd.PrimaryXmlWriter`.
+
+
+### Metadata and signature lookup
+
+In order to verify integrity of the `Packages.gz` file Uyuni must first lookup the `Release` file and verify the checksum it contains and its GPG signature. This file can be in one of these locations:
+- `http://mirror.example.com/$ARCHIVE_ROOT/dists/$DIST` in case of repos with a `dists` directory when the Uyuni repo URL points to `http://mirror.example.com/$ARCHIVE_ROOT/dists/$DIST/$COMPONENT/binary-$ARCH/`.
+- `http://mirror.example.com/$REPODIR/` in the case of flat repositories
+
+Looking up the `Release` file should be done automatically when syncing of the repo. An error should be shown if it's not possible to locate the `Release` file and its corresponding signature.
+
+### Metadata Checksum Verification
+
+Uyuni should verify the integrity of the `Packages.gz|xz` files using the checksums present in the `Release` file.
+The integrity of the `deb` binary packages must be verified using the checksums from the `Packages` file.
+
+### GPG signature checking and checksum verification
+
+This will be handled in a separate RFC.
+
+## Failure Registry
+
+Packages might have sync issues, such as wrong GPG signature or
+checksum mismatch or downloading issues etc. However, broken packages
```

Comment:
As GPG signature checking and checksum verification is handled in a [separate RFC](https://github.com/uyuni-project/uyuni-rfc/pull/12), wouldn't the corresponding error handling better be described there as well?

Reply:
If reposync would be a 15 minutes deal, I would rather indeed discarding the entire batch. But reposync can hang there for long hours. So you want to try to fix few files (and if that still failed, e.g. remote service temporarily out of serivice, hence the broken last two packages or so) you just may come back later and fix only those two packages.

If you want to delete the whole thing, because you gave up on that broken repo — also fine, because we've got the whole list of those files and we know where they are located on the disk.

Reply:
Personally I would feel better that UX provides some sort of "try again" fix button for only those broken files alongside "Delete repo". So then as an admin, I can still poke it couple of times and decide I want to keep it broken, wait for repo to be fixed or delete it entirely. If we discard it on error, it is fine for us, but then the UX experience a bit annoying: you never know if the other side will sync or not.

Basically, I am only proposing here to build here `rsync` analogy that resumes broken download (and we do know it is broken!)  instead of `scp` that restarts all over again, even only two files left to resume out of thousand.

@paususe @mateiw @mcalmer thoughts?

Reply:
> Personally I would feel better that UX provides some sort of "try again" fix button for only those broken files alongside "Delete repo". So then as an admin, I can still poke it couple of times and decide I want to keep it broken, wait for repo to be fixed or delete it entirely. If we discard it on error, it is fine for us, but then the UX experience a bit annoying: you never know if the other side will sync or not, yet again wait for hours it complete (or, again, not).
> 
> Basically, I am only proposing here to build here `rsync` analogy that resumes broken download (and we do know it is broken!) instead of `scp` that restarts all over again, even only two files left to resume out of thousand.
> 
> @paususe @mateiw @mcalmer thoughts?

Agreed. 

@joesusecom even mentioned (can't remember where) that we should automatically retry, at least once, missing/broken packages after the reposync has ended.

Reply:
Well, retry doesn't necessarily helps. It might if the network or storage glitch is very short or local. But that is very rare, apparently. More often happens that the mirror storage (or mounted NFS glitched somewhere) causing broken package, then just another retry won't help here. We need that retry delayed for much longer period.

Reply:
Plus another (sorry for writings) reason to _keep_ the broken repo: say, `vim` package is broken. But you want to install `emacs`. That mean you still can perfectly use that broken channel, even so we know that it is broken.